src/main/java/com/google/cloud/spanner/r2dbc/client/GrpcClient.java

```
@@ -191,56 +198,40 @@ public GrpcClient(GoogleCredentials credentials) {
   }
 
   @Override
-  public Mono<ExecuteBatchDmlResponse> executeBatchDml(StatementExecutionContext ctx, String sql,
-      List<Struct> params, Map<String, Type> types) {
-    return Mono.defer(() -> {
-      ExecuteBatchDmlRequest.Builder request = createBatchDmlRequestBuilder(ctx);
-      for (Struct paramsStruct : params) {
-        ExecuteBatchDmlRequest.Statement statement = ExecuteBatchDmlRequest.Statement.newBuilder()
-            .setSql(sql).setParams(paramsStruct).putAllParamTypes(types)
-            .build();
-        request.addStatements(statement);
-      }
-
-      return ObservableReactiveUtil
-          .unaryCall(obs -> this.spanner.executeBatchDml(request.build(), obs));
-    });
-  }
+  public Mono<ExecuteBatchDmlResponse> executeBatchDml(
+      StatementExecutionContext ctx,
+      List<Statement> statements) {
 
-  @Override
-  public Mono<ExecuteBatchDmlResponse> executeBatchDml(StatementExecutionContext ctx,
-      List<String> statements) {
     return Mono.defer(() -> {
-      ExecuteBatchDmlRequest.Builder request = createBatchDmlRequestBuilder(ctx);
-      for (String sql : statements) {
-        ExecuteBatchDmlRequest.Statement statement = ExecuteBatchDmlRequest.Statement.newBuilder()
-            .setSql(sql)
-            .build();
-        request.addStatements(statement);
-      }
+      ExecuteBatchDmlRequest.Builder request =
+          ExecuteBatchDmlRequest.newBuilder()
+              .setSession(ctx.getSessionName())
+              .addAllStatements(statements)
+              .setSeqno(ctx.nextSeqNum());
 
-      return ObservableReactiveUtil
-          .unaryCall(obs -> this.spanner.executeBatchDml(request.build(), obs));
+      if (ctx.getTransactionId() != null) {
```

Hmmm there's a difference here it seems.

So the `ctx.getTransactionId() != null` here happens inside the mono.defer of the response, which means this context isn't checked until after the user tries to consume the result mono. 

Previously we were checking this condition on creating the mono (not retrieving the results from the mono). 

I'm not quite sure which one is correct, but that's definitely a difference.

---

Are you sure that is correct? I think it may be the same. In the previous code, the `ctx.getTransactionId() != null` check was done in the `createBatchDmlRequestBuilder()` method; that lives inside the Mono.defer.

---

I was looking at the one in `AutoCommitStatement`. That one is done when statement.execute() is called, which is outside of the mono.defer

---

I see. In that case I think the current way is preferable because the statement no longer needs to know about the transaction state.

In the previous way, we could probably construct a case where a user builds a full stream (connection.createStatement().execute()) when a transaction is active but then subscribe to it after the transaction is committed; this might be another case which breaks the previous impl.

---

Actually I think this is an illustration why it is worse.

I think the user's intention is defined by how he writes the code that creates the mono. In other words, if the user calls "execute" on the statement between "beginTransaction" and "commitTransaction" that means he wants this statement to execute inside the transaction (not as auto-commit). This .execute call is BEFORE the mono is consumed, so we need to know the ctx state at the mono creation time. 

If the user then consumes this execute result later then according to your current code it would execute as a stand-alone , which is NOT what he intends in my opinion. He should be told "you tried to do this DML inside a transaction, but in fact that transaction has ended and you tried consuming it too late."

---

Hmm, I see your point. Perhaps this discussion is ultimately moot because the user will be using the pattern of chaining their `begin/execute/commit` Monos together; this edge-case we are discussing is contrived.

Overall I think it is the reactive paradigm to do all your work inside a Mono.defer and not immediately, at least this seemed to be the heuristic/style used, so I am still inclined to keep it this way.

---

But isn't it actually kinda easy to do this?

```

factory.create()
.delayUntil(c->c.beginTransaction)
.flatMapMany( c-> c.createStatement("SELECT accounts to transfer here..." ).execute())
.map(c-> c.createSTatement( [delete money from account here]).execute())
.delayUntil(c-> c.commitTransaction())
.map(c-> print("let's see how much money we just deleted: " + c.getRowsUpdated());

// right there you would be deleting money from accounts non-atomically in a separate condition just because you don't check the results until after the commit 
```

And while it's easy to make a blanket statement to say they prefer doing everything in a mono.defer, there's the implicit "if it doesn't change the actual logic" at the end!

---

Hmm, ok, in that case it should return an error right? Instead of run? Yeah I see your point.
I would want to ask mark to understand what he intended for this interface just to confirm. The mental block for me still is that I feel that the statement should not need to know about transaction state.

Also note this example actually is protected against in our driver. I think due to idiosyncrasy of Spanner we had to make it so DML queries will actually execute on `execute()`; however select queries will defer the results are retrieved.

---

Hey @mp911de -

We have a question about the behavior of `Statement.execute()`

In the below example, the user starts a transaction to execute a *read* and then *write* on a table. However, the user by mistake calls `c.getRowsUpdated()` outside the `commitTransaction()` on the second statement. As we understand it, the user must get the results of a query for it to be fully run in Reactive; thus, the user would execute the statement outside the transaction in the below example.

```
factory.create()
.delayUntil(c->c.beginTransaction)
.flatMapMany( c-> c.createStatement("SELECT accounts to transfer here..." ).execute())
.map(c-> c.createSTatement( [delete money from account here]).execute())
.delayUntil(c-> c.commitTransaction())
.map(c-> print("let's see how much money we just deleted: " + c.getRowsUpdated());

// right there you would be deleting money from accounts non-atomically in a separate condition just because you don't check the results until after the commit
```

In this case, when the statement is run: Should the statement be run as a stand-alone, autocommit statement? Or should an error be thrown that the transaction no longer exists?

---

@dzou that behavior is not defined. What typically happens is that  `.execute()` participates in the transaction and you get a `Result` object that is attached to the transaction. Depending on whether you have all results in memory, or whether you need to fetch data, behavior can be different. If you have all results in memory and you call `getRowsUpdated()` after committing the transaction, then you could return the cached data.

If you need to issue another query/talk to the server, after the transaction is committed, then probably you would run into an error.

It's fair to say that any object associated with a transaction gets invalidated if the transaction gets committed. This is usual for SQL databases (e.g. close a result set when the transaction gets cleaned up or invalidate BLOB/CLOB objects after a transaction).

Does this help?

---

Thanks Mark, that is helpful.

@ChengyuanZhao - So to resolve this discussion, would it be reasonable to just move forward with this PR and then create an issue to revisit this edge case later? The reason for deferring this would be that 1. our driver behaves correctly because we run DML queries on `execute()` 2. I would need to think about what are the correct changes to introduce to make the SpannerStatement track the transaction state.

---

sounds good to me!

---

Ok, created: #150.